core/src/subgraph/runner.rs

```diff
@@ -556,6 +574,81 @@ where
 
         Ok(action)
     }
+
+    async fn handle_offchain_queue(&mut self) -> Result<(), Error> {
+        loop {
+            if let Ok((cid, data)) = self.ctx.offchain_monitor.ipfs_monitor_rx.try_recv() {
+                self.handle_offchain_update(offchain::Source::Ipfs(cid), data)
+                    .await?;
+                continue;
+            }
+            break;
+        }
+        Ok(())
+    }
+
+    async fn handle_offchain_update(
+        &mut self,
+        source: offchain::Source,
+        data: bytes::Bytes,
+    ) -> Result<(), Error> {
+        let matching_data_sources = self
+            .ctx
+            .offchain_monitor
+            .data_sources
+            .iter()
+            .filter(|ds| ds.source.as_ref() == Some(&source));
+
+        let mut block_state =
+            BlockState::<C>::new(self.inputs.store.cheap_clone(), LfuCache::new());
+        let entity_cache = std::mem::take(&mut self.state.entity_lfu_cache);
+
+        self.persist_dynamic_data_sources(
+            &mut block_state.entity_cache,
+            matching_data_sources
+                .cloned()
+                .map(data_source::DataSource::Offchain)
+                .collect(),
+        );
+
+        let causality_region = match &source {
+            offchain::Source::Ipfs(cid) => format!("ipfs/{}", cid.to_string()),
+        };
+        let trigger = TriggerData::Offchain(offchain::TriggerData {
+            source: source.clone(),
+            data: Arc::new(data),
+        });
+
+        let block_state = self
+            .ctx
+            .instance
+            .process_trigger(
+                &self.logger,
+                &Arc::default(),
```

Comment:
This is not great, though it does seem to meet our initial requirements. Worth a comment that this will need further refactoring at some point.